neverbleed.c

```diff
@@ -568,23 +778,47 @@ int neverbleed_load_private_key_file(neverbleed_t *nb, SSL_CTX *ctx, const char
 
     if (expbuf_read(&buf, thdata->fd) != 0)
         dief(errno != 0 ? "read error" : "connection closed by daemon");
-    if (expbuf_shift_num(&buf, &ret) != 0 || expbuf_shift_num(&buf, &key_index) != 0 || (estr = expbuf_shift_str(&buf)) == NULL ||
-        (nstr = expbuf_shift_str(&buf)) == NULL || (errstr = expbuf_shift_str(&buf)) == NULL) {
+    if (expbuf_shift_num(&buf, &ret) != 0 || expbuf_shift_num(&buf, &type) != 0 || expbuf_shift_num(&buf, &key_index) != 0) {
         errno = 0;
         dief("failed to parse response");
     }
 
-    if (ret != 1) {
+    switch (type) {
+    case NEVERBLEED_TYPE_RSA:
+        if ((estr = expbuf_shift_str(&buf)) == NULL || (nstr = expbuf_shift_str(&buf)) == NULL) {
+            errno = 0;
+            dief("failed to parse response");
+        }
+        pkey = create_pkey(nb, key_index, estr, nstr);
+        break;
+#if OPENSSL_101_API
+    case NEVERBLEED_TYPE_ECDSA:
+        if (expbuf_shift_num(&buf, &curve_name) != 0 || (ec_pubkeystr = expbuf_shift_str(&buf)) == NULL) {
+            errno = 0;
+            dief("failed to parse response");
+        }
+        pkey = ecdsa_create_pkey(nb, key_index, curve_name, ec_pubkeystr);
+        break;
+#endif
+    default:
+        if ((errstr = expbuf_shift_str(&buf)) == NULL) {
```

Comment:
I assume we should just return an error to the peer without trying to parse arguments calling by expbuf_shift_xxx, considering the fact that we cannot determine the number and types of the arguments unless the type is TYPE_RSA or TYPE_ECDSA.